### ISSUE TITLE
[Feature] pnpm/corepack support

### ISSUE THREAD

**Issue by dtinth** [microsoft/Oryx#2340]
### Feature Request

From https://github.com/microsoft/Oryx/issues/1150#issuecomment-1886067023

> We're not _currently_ looking to support pnpm. Closing this issue for now, please create a new issue and link to this one if you feel strongly.

I feel strongly about this and decided to open a new issue.

My suggestions is at follows:

- Detect presence of `packageManager` key in `package.json`.
- If found,
   - Run `corepack enable`
   - Run `npm install` / `yarn install` / `pnpm install` depending on the package manager name
- If not found,
   - Just run `npm install` as before

How this will improve things:

- Azure Static Site and Azure Web Service users can build their app without having to set up `PRE_BUILD_COMMAND` and `CUSTOM_BUILD_COMMAND`. I use Netlify and Vercel, and they are supported out-of-the-box, so I expect the same in other hosting services. Azure services not recognizing pnpm in my project is a turn off for me.

- GitHub Codespaces currently run `npm install` even in pnpm projects. This means not only it takes a longer time for Codespace to initialize pnpm projects (due to waiting for npm install), afterwards I have to go ahead and `rm -rf node_modules package-lock.json` and run `pnpm install` again.

In short, supporting pnpm via corepack will improve both developer experience in Codespaces and deployment experiences in Azure services.

**Comment by simonjj** [comment]
Thank you for opening this issue and linking back.

**Comment by PhantipK** [comment]
Please update and scale the Oryx to support for pnpm, so it can reach a broader audience including OSS developers who switched to pnpm and businesses using Azure Static Web Apps.

**Comment by GalindoSVQ** [comment]
What about using [pnpm/action-setup@v3](https://github.com/pnpm/action-setup) ?

**Comment by fturmel** [comment]
@GalindoSVQ From my experience deploying a NextJS app last week, these PRE_BUILD_COMMAND and CUSTOM_BUILD_COMMAND variables did work to some extent.

The blocker was that during the Azure Functions build phase, `npm install --production` would always run regardless and I could not find a way to override that part of the pipeline. I was forced to alter my app deps to use npm to successfully deploy.

**Comment by GalindoSVQ** [comment]
> @GalindoSVQ From my experience deploying a NextJS app last week, these PRE_BUILD_COMMAND and CUSTOM_BUILD_COMMAND variables did work to some extent.
> 
> The blocker was that during the Azure Functions build phase, `npm install --production` would always run regardless and I could not find a way to override that part of the pipeline. I was forced to alter my app deps to use npm to successfully deploy.

react app boostraped using vite works fine, I am not sure but.. have you tried specifying pnpm as engine in package.json? 

https://docs.npmjs.com/cli/v10/configuring-npm/package-json#engines

**Comment by fturmel** [comment]
> react app boostraped using vite works fine

That makes sense to me, it's probably a static app that doesn't have APIs/backend and doesn't use Azure Functions? Or maybe your project isn't pruning dev dependencies and you're not hitting the same build/deploy path I am.

I don't think specifying pnpm in package.json will make a difference. Everything in Oryx seems hardcoded to use npm or yarn exclusively.

I can track the problem here for my specific blocker (in case it helps the maintainers in any way), where `ProductionOnlyPackageInstallCommand ` always runs and there are no overrides to make `packageInstallCommand` be `pnpm install`.

https://github.com/microsoft/Oryx/blob/8fb59f3713371b0d39b4b4df3a96b3a36b9c52a7/src/BuildScriptGenerator/Node/NodeBashBuildSnippet.sh.tpl#L120-L125

https://github.com/microsoft/Oryx/blob/8fb59f3713371b0d39b4b4df3a96b3a36b9c52a7/src/BuildScriptGenerator/Node/NodePlatform.cs#L196-L218

**Comment by asheemak** [comment]
same issue here cannot use pnpm

**Comment by faxg** [comment]
+1

**Comment by itpropro** [comment]
Guys, please update Oryx, so it can be used on modern TS/JS projects.
Not supporting pnpm, bun and yarn in 2025 is a little absurd.

**Comment by PhantipK** [comment]
Any updates on this?

**Comment by Syntarex** [comment]
Would love to see this too. ♥️

**Comment by carlohoanguyen** [comment]
+1!

**Comment by TheAkshitS** [comment]
+1

**Comment by obada-jaras** [comment]
waiting for `bun` support

**Comment by awensaunders** [comment]
I'd like to echo that this is important. `npm` allows postInstall scripts by default, and as such many organisations are moving to `pnpm` as a drop-in replacement with significantly superior security controls against supply chain attacks. `pnpm` is rapidly becoming the package manager of choice for new projects. It should also be possible to install the managed api with a custom package manager such as `pnpm`, which the proposed workarounds do not allow as far as I understand.
